packet/packet.go

```diff
@@ -213,6 +226,53 @@ func (packet *Packet) ParseL3() {
 	packet.L3 = unsafe.Pointer(packet.unparsed())
 }
 
+// Correctly returns EtherType from Ethernet header or VLAN header.
+func (packet *Packet) GetEtherType() uint16 {
+	if packet.Ether.EtherType == SwapBytesUint16(VLANNumber) {
+		vptr := uintptr(unsafe.Pointer(packet.Ether)) + EtherLen
+		vhdr := (*VLANHdr)(unsafe.Pointer(vptr))
+		return vhdr.EtherType
+	} else {
+		return packet.Ether.EtherType
+	}
+}
+
+// ParseL3 set poinetr to start of L3 header taking possible presence
+// of VLAN header into account.
+func (packet *Packet) ParseL3CheckVLAN() {
+	ptr := uintptr(unsafe.Pointer(packet.Ether)) + EtherLen
+
+	if packet.Ether.EtherType == SwapBytesUint16(VLANNumber) {
+		ptr += VLANLen
+	}
+	packet.L3 = unsafe.Pointer(ptr)
+}
+
+// ParseL3 set poinetr to start of L3 header taking possible VLAN
+// presence into account. It returns VLAN header if present and always
+// correct EtherType from VLAN header or Ethernet header.
+func (packet *Packet) ParseVLANAndL3() (uint16, *VLANHdr) {
+	ptr := uintptr(unsafe.Pointer(packet.Ether)) + EtherLen
+
+	if packet.Ether.EtherType == SwapBytesUint16(VLANNumber) {
+		vhdr := (*VLANHdr)(unsafe.Pointer(ptr))
+		ptr += VLANLen
+		packet.L3 = unsafe.Pointer(ptr)
+		return vhdr.EtherType, vhdr
+	} else {
+		packet.L3 = unsafe.Pointer(ptr)
+		return packet.Ether.EtherType, nil
+	}
+}
+
+// GetVLAN returns VLAN header pointer if it is present in the packet.
+func (packet *Packet) GetVLAN() *VLANHdr {
+	if packet.Ether.EtherType == SwapBytesUint16(VLANNumber) {
+		return (*VLANHdr)(unsafe.Pointer(uintptr(unsafe.Pointer(packet.Ether)) + EtherLen))
```

Comment:
packet.Ether + EtherLen **-2** - no? VLAN tag not after etherHdr, but inside it, before etherType which is 2 byte long